Source/inv.cpp

```diff
@@ -704,6 +708,65 @@ bool AutoPlaceItemInInventory(int playerNumber, const ItemStruct &item, bool per
 	return done;
 }
 
+/**
+ * @brief Checks whether the given item can be placed on the specified player's inventory.
+ * If 'persistItem' is 'True', the item is also placed in the inventory, using an alternative
+ * algorithm that should typically provide better packing than AutoPlaceItemInInventory().
+ * @param playerNumber The player number on whose inventory will be checked.
+ * @param item The item to be checked.
+ * @param persistItem Pass 'True' to actually place the item in the inventory. The default is 'False'.
+ * @return 'True' in case the item can be placed on the player's inventory and 'False' otherwise.
+ */
+bool AutoPlaceItemInInventoryAlternate(int playerNumber, const ItemStruct &item, bool persistItem)
+{
+	InvXY itemSize = GetInventorySize(item);
+	bool done = false;
+
+	if (itemSize.Y == 1) {
+		for (int i = 30; i <= 39 && !done; i++) {
+			done = AutoPlaceItemInInventorySlot(playerNumber, i, item, persistItem);
+		}
+		for (int x = 9; x >= 0 && !done; x--) {
+			for (int y = 2; y >= 0 && !done; y--) {
+				done = AutoPlaceItemInInventorySlot(playerNumber, 10 * y + x, item, persistItem);
+			}
+		}
+	}
+
+	if (itemSize.Y == 2) {
+		for (int x = 10 - itemSize.X; x >= 0 && !done; x -= itemSize.Y) {

```

Comment:
item y is known to be 2, probably should have been x. but also jumping a full item width may skip possibly good spots

```
xoox
xoox
```
A book can fit in the middle but if i move by the size of a book both options are half filled.

```suggestion
		for (int x = 10 - itemSize.X; x >= 0 && !done; x--) {
```

Reply:
Yes, that should have been X, thanks.

As to why it's `x -= itemSize.X`: I did actually write it as just `x--` the first time but I played with that for half an hour or so and decided I didn't like the result.

The original placement algorithm for 2x2 items starts out by trying a set list of preferred positions before falling back to searching one square at a time, and without that it's very easy to end up with two rows of books that are misaligned by one. I don't *think* that misalignment ever results in technically worse packing, but I also don't think it results in technically *better* packing either (because you just end up with a 1x1 or 1x2 gap in a different place), so I decided to go with the version that looks neater visually.

This does lead to a bit of special casing for 2x2 items, so I didn't quite succeed in my original plan to make 2x2 and 1x2 items use the same placement, but it does at least have less special-casing than the original.